WebPageProxy::setToolTip accesses a null PageClient

aprotyas opened the pull request:
#### 46b03afa5a231970ac09fdd9cbfeeaaa68dc31b0
<pre>
WebPageProxy::setToolTip accesses a null PageClient
<a href="https://bugs.webkit.org/show_bug.cgi?id=258313">https://bugs.webkit.org/show_bug.cgi?id=258313</a>
rdar://110501313

Reviewed by Chris Dumez and Wenson Hsieh.

In 264455@main, we introduced
`WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously`, which asks
to perform `WebPageProxy::mouseDidMoveOverElement` at a future time. It
turns out there&apos;s a null access on `m_pageClient` when we then call into
`WebPageProxy::setToolTip`, which happens because we hold a
`WeakPtr&lt;PageClient&gt;` instance to the `PageClient` object, meaning there
are no guarantees on the PageClient instance&apos;s lifetime.

To avoid this null dereference, we should only inform the web page about
a `MouseMove` event upon consulting whether the web page is closed
through `WebPageProxy::isClosed()`.

We also add an API test that crashes without this mitigation in place.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(+[TestEventMonitor addLocalMonitorForEventsMatchingMask:handler:]):

Canonical link: <a href="https://commits.webkit.org/265456@main">https://commits.webkit.org/265456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b9bcceb47a78b4416bc5cfd9e869284d25e0e16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10911 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12558 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13501 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13353 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12962 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9856 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17090 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10335 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10008 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13253 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10449 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8540 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9624 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2627 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13891 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10309 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6970,7 +6970,8 @@ void WebPageProxy::mouseDidMoveOverElement(WebHitTestResultData&& hitTestResultD
     m_lastMouseMoveHitTestResult = API::HitTestResult::create(hitTestResultData);
     m_uiClient->mouseDidMoveOverElement(*this, hitTestResultData, modifiers, m_process->transformHandlesToObjects(userData.object()).get());
-    setToolTip(hitTestResultData.toolTipText);
+    if (m_pageClient)
```
pageClient is not meant to be null checked, which is why pageClient() returns a reference. It would be better to make sure the page is not closed (do a `WebPageProxy::isClosed()` check).

cdumez replied:
Also, we may want to do this check earlier. Probably not much point in doing the earlier work if the page is closed.

aprotyas replied:
> pageClient is not meant to be null checked, which is why pageClient() returns a reference

Yeah it did seem sort of an anti-pattern to do so inside `WebPageProxy`

aprotyas replied:
> It would be better to make sure the page is not closed (do a WebPageProxy::isClosed() check).

The reason we hit this issue is because we're async dispatching a completion handler that calls `WebPageProxy::mouseDidMoveOverElement()`, so we can't consult `WebPageProxy::isClosed()` when that work is scheduled.

Thoughts?

aprotyas replied:
Hmmm, maybe the completion handler itself could consult `WebPageProxy::isClosed()` before calling into `WebPageProxy::mouseDidMoveOverElement()`

cdumez replied:
Either early return in this function or before calling this function in the lambda. I don’t feel strongly either way. However, it should be a IsClosed() check.

aprotyas replied:
Thanks for the catch again, fixed in https://github.com/WebKit/WebKit/commit/3547821b6b8c0933194eb6bb9ec328c6bfd166ae

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3204,7 +3204,8 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
 {
     sendWithAsyncReply(Messages::WebPage::PerformHitTestForMouseEvent { event }, [this, protectedThis = Ref { *this }](WebHitTestResultData&& hitTestResult, OptionSet<WebEventModifier> modifiers, UserData&& userData) {
-        this->mouseDidMoveOverElement(WTFMove(hitTestResult), modifiers, WTFMove(userData));
+        if (!this->isClosed())
```
Do we really need `this->`?

aprotyas replied:
No we don't! Removed `this->` for better readability in https://github.com/WebKit/WebKit/commit/2b9bcceb47a78b4416bc5cfd9e869284d25e0e16

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3204,7 +3204,8 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
     sendWithAsyncReply(Messages::WebPage::PerformHitTestForMouseEvent { event }, [this, protectedThis = Ref { *this }](WebHitTestResultData&& hitTestResult, OptionSet<WebEventModifier> modifiers, UserData&& userData) {
-        this->mouseDidMoveOverElement(WTFMove(hitTestResult), modifiers, WTFMove(userData));
+        if (!this->isClosed())
+            this->mouseDidMoveOverElement(WTFMove(hitTestResult), modifiers, WTFMove(userData));
```
ditto.

aprotyas replied:
https://github.com/WebKit/WebKit/commit/2b9bcceb47a78b4416bc5cfd9e869284d25e0e16

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
```diff
@@ -979,6 +981,56 @@ - (void)_webView:(WKWebView *)webview mouseDidMoveOverElement:(_WKHitTestResult
+    }));
+
+    @autoreleasepool {
+        WKWebViewConfiguration *configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"FrameHandleSerialization"];
+        auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration]);
```
~Nit - I think this `WKWebViewConfiguration` (which initializes the "FrameHandleSerialization" plugin) is also unnecessary.~
```suggestion
        auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300)]);
```

whsieh replied:
(Never mind, looks like it's necessary because of `MouseMoveOverElementDelegate`)

aprotyas replied:
Yeah, the `MouseMoveOverElementDelegate` in this test file hardcodes a userInfo class check against `_WKFrameHandle`, so we either use this configuration or add a workaround. Staying with this for now then

aprotyas committed:
WebPageProxy::setToolTip accesses a null PageClient
https://bugs.webkit.org/show_bug.cgi?id=258313
rdar://110501313

Reviewed by Chris Dumez and Wenson Hsieh.

In 264455@main, we introduced
`WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously`, which asks
to perform `WebPageProxy::mouseDidMoveOverElement` at a future time. It
turns out there's a null access on `m_pageClient` when we then call into
`WebPageProxy::setToolTip`, which happens because we hold a
`WeakPtr<PageClient>` instance to the `PageClient` object, meaning there
are no guarantees on the PageClient instance's lifetime.

To avoid this null dereference, we should only inform the web page about
a `MouseMove` event upon consulting whether the web page is closed
through `WebPageProxy::isClosed()`.

We also add an API test that crashes without this mitigation in place.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(+[TestEventMonitor addLocalMonitorForEventsMatchingMask:handler:]):

Canonical link: https://commits.webkit.org/265456@main
